feat: [USER-SIGN-IN] "add sign-in user & token"

jnjeaaaat opened the pull request:
Changes
---
1. SignController, SignService
   - 유저 로그인 기능을 위해 api추가
   - 아이디, 비밀번호 를 받아서 유효성 처리
   - 로그인 하면 `accessToken`, `refreshToken` 발급
2. AccessToken, RefreshToken Entity Class 생성
   - `RedisHash` 어노테이션으로 `RedisServer에서 관리`할 수 있게 함
3. 유저 정보를 받아 토큰을 생성, 조회 할 `JwtUtil` 생성
   - accessToken 만료기간은 30분, refreshToken 만료기간은 2주로 잡고 토큰 생성
   - 유저 권한(role-list) 확인 가능
   - 헤더로 넘어온 accessToken값 추출 가능
4. `JwtFilter` 추가
   - Jwt Token 을 관리할 Filter 생성
   - Filter에서 발생하는 토큰 예외 처리
5. SpringSecurity 에 필터추가, `필터 예외처리`를 할 Handler 추가
   - SpringSecurity 에서 `antMatcher().hasRole()` 메소드로 권한 분리를 해줘야 `ExceptionHandler가 적용`된다.

Background
---
1. 로그인 기능을 구현하기 위해 기본 틀 생성, 토큰 발행 후 기능들에 대한 유저 접근 권한 관리를 해야한다.
2. 토큰은 만료기간이 존재하기 때문에 RedisServer에서 관리해야한다.
3. 토큰을 확인하고 거르는 역할을 해줄 Filter가 필요하다.
4. Filter에 대한 예외처리는 ControllerAdvice에서 불가능하기 때문에 따로 예외처리를 해줘야한다.
5. SecurityConfig에 권한에 따른 api를 설정해줘야한다.

Discuss
---
accessToken, refreshToken을 둘다 관리해서 로그인 기능을 구현해본적이 없어서 힘들었지만,
열심히 코드를 조합해서 기능 구현에 성공했습니다....

Filter에서 try,catch 문이 중첩되어있어서 깔끔해보이지 않지만
어떻게 더이상 더 나은코드로 바꿔야할지 감이 잡히질 않습니다..

지금은 일단 성공적으로 기능이 동작하는 것에 만족해야될 것 같습니다

Issues
---
Related to: #19 #20 #21

jnjeaaaat committed:
feat: [USER-SIGNIN] "add sign-in user"

아이디, 비밀번호 으로 로그인 요청

Related to: #19

jnjeaaaat committed:
feat: [USER-SIGNIN] "add token service"

로그인 하면 accessToken, refreshToken 발행 후 Redis Server에 저장, 조회, 삭제 관리
토큰 처리를 위해 JwtFilter 추가

Related to: #20

jnjeaaaat committed:
feat: [USER-SIGNIN] "handle token exception"

토큰 유효성(만료, null, 존재하지 않는 토큰) 문제, 권한이 없는 유저의 접근에 대한 Exception 처리

Related to: #21

mingming-mentor commented on src/main/java/org/jnjeaaaat/onbition/config/filter/CustomAccessDeniedHandler.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public void handle(HttpServletRequest request, HttpServletResponse response, AccessDeniedException accessDeniedException) throws IOException, ServletException {
+        log.error("[handle] 접근이 막혔을 경우 에러 메세지 리턴");
+        ObjectMapper objectMapper = new ObjectMapper();
```
bean으로 만들어서 사용해주시면 좋을 것 같아요.

jnjeaaaat replied:
넵 수정했습니다😃

mingming-mentor commented on src/main/java/org/jnjeaaaat/onbition/config/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,118 @@
+        SecurityContextHolder.getContext().setAuthentication(authentication); // SpringSecurity 에 권한 목록 넘겨줌
+        log.info("[doFilterInternal] token 값 유효성 체크 완료");
+      }
+    } catch (ExpiredJwtException e) { // accessToken 이 만료되었을때
```
accessToken이 만료되면 자동으로 refreshToken을 꺼내와서 accessToken을 만들어주는 형태인가요?
API request에서 refreshToken을 보내줄 때 accessToken을 발급해줘야하지 않나 싶어요.
이러면 accessToken이 만료가 되어 더이상 유효하지 않을 때에도 accessToken만 탈취한다면 새로운 accessToken을 발급받을 수 있어보이네요

jnjeaaaat replied:
피드백 반영하여 수정했습니다.😃

mingming-mentor reviewed (approved):
리뷰드렸습니다. 수고 많으셨어요

jnjeaaaat committed:
feat: "ObjectMapper bean 등록"

ObjectMapper 를 생성자를 통해 받아오지 않고 bean 으로 등록

jnjeaaaat committed:
refactor: "#22 feedback 반영"

accessToken을 탈취당한다면 만료된 토큰으로 계속 새로운 accessToken을 요청하는 것을 방지하기 위한 '재발급 api' 추가

Resolved: #23

jnjeaaaat committed:
feat: "add logout api"

accessToken을 받아 logout 처리

Resolved: #24

nadoran78 commented:
저도 access, refresh 토큰을 구성해보는 게 처음이라 고민이었는데, 쓰신 코드 보고 방향성을 알게 되었습니다. 감사합니다.

jnjeaaaat committed:
test: "test login"

login 처리 성공, 실패 테스트코드 작성

Resolved: #25
